### PR TITLE
// [+] Product Search with Facets

### PR DESCRIPTION
**Pull request by djfm**
We want to standardize the way modules can hook into the product listing front-controllers to manage product search and filters.

This should make creating faceted search modules easier and safer. We demonstrate this by adapting blocklayered to this new infrastructure, so please see [blocklayered](https://github.com/djfm/blocklayered/tree/feat/starter-theme) together with this PR.

For more information, please read the FACETED_SEARCH.md file and the extensive comments in `ProductListingFrontControllerCore`.

### CODE REVIEW

**Comment by inetbiz** [comment]
@djfm :+1:  :bow:

File: themes/StarterTheme/templates/catalog/products.tpl [hunk 1]
```diff
@@ -1,17 +1,21 @@
-{if $products|count}
-  <section id="products">
-    <h1>{l s='Products'}</h1>
+<section id="products">
+  {if $products|count}
```
Comment (inetbiz) [inline themes/StarterTheme/templates/catalog/products.tpl]: Still have to get products into a list for the schema on a category listing. [@mfhepp|Dr. Martin Hepp](https://github.com/mfhepp) of good relations explains it best on https://github.com/schemaorg/schemaorg/issues/605

**Comment by kpodemski** [comment]
@djfm don't you think that some classes aren't valid with SOLID design principles? I think shouldn't we have this:
https://github.com/djfm/PrestaShop/blob/feat/starter-theme/Core/Business/Product/Search/ProductSearchQuery.php

as an interface used by parent class for each of entity?

**Comment by inetbiz** [comment]
Mentioned PR at https://goo.gl/cCagqy

**Comment by djfm** [comment]
@kpodemski You mean a Query interface and then a class for Category, Manufacturer, Supplier, etc.? I hesitated, I was afraid it would *seem* like too many classes. To be really SOLID I should have a CategoryFilter, AttributeFilter, etc. too

**Comment by djfm** [comment]
@inetbiz not sure I'm following. Do you mean to add properties like "color: red", "size: s" in the listing? Does it need to be linked with the facets? Or would adding the attributes in a schema.org compliant way on product-miniature.tpl (regardless of facets) do the trick?

**Comment by kpodemski** [comment]
@djfm well, nvm, SOLID is not used in most of PrestaShop classes so it was wrong idea

about new search mechanism, well done, for blocklayered main changes for sure must be:
- ability to display filters on pages like "New products", "Prices drop", "Search results"
- ability to filter product "On sale", "New"
- possibility to extend filters by our own

**Comment by djfm** [comment]
@kpodemski 

blocklayered is **very hard** to add features to, so the features you describe won't be in blocklayered itself, which was only adapted with the same level of features (minus a few options for simplification, plus a few bugfixes, like, now the price and weight filters take the other filters into account)

But with the ProductListingFrontController re-architecture I propose, it becomes much easier to develop faceted search modules that work on all pages. We have yet to figure something out for the search too, which wasn't changed at all in this PR.

> possibility to extend filters by our own

That's why the `Facet` and `Filter` classes are generic and not specific for each type of filter: your modules can return any kind of filters they want so it is easy to add new simple filters (at the moment the core supports only radio buttons and checkboxes, which is controlled by the $multipleSelectionAllowed field of Facet ($multipleSelectionAllowed === true => checkbox, radio otherwise))

If a module needs to render more advanced filters then it can do so by returning an object from the`productSearchProvider` hook that implements `FacetsRendererInterface` in addition to `ProductSearchProviderInterface`. Then the module's filters will be rendered by the module itself and anything is possible.

Also, I don't know if you've noted, but the faceted search modules don't need to do all the hard work of fetching all product information, they can just return a list of id_product's and the core will add what's necessary in a consistent way.

**Comment by kpodemski** [comment]
@djfm 

Ok, I got it :) 

I'll play with this after it will be merged, I'll try to create some additional filters using `productSearchProvider`, I'm happy that you get rid off those "getProductProperties" everywhere!

**Comment by djfm** [comment]
@kpodemski preferably play with it before it is merged if you can, as any problem you can find before the merge is that less noise in the official repo :)

I'm wondering if I should keep `hookActionProductListOverride` in case no search providers were found or just get rid of it. Any opinion?

**Comment by kpodemski** [comment]
@djfm keep it, I used it for example to retrive all product images to create gallery on product listing, possibility to retrive all product attributes/features and allow to quick order product from product listing and few more things, of course it can be done by overrides but it was nice to play with those features without creating them however...

 I'm thinking that maybe you can create standarized methods for this? some method to return formatted combinations like i.e product page and formatted images like on product page

but one way or another modifying product list dynamically from module is super useful for many reasons

**Comment by djfm** [comment]
@kpodemski Point taken, makes sense. So in this scenario, `hookActionProductListOverride` would be executed **after** the products are fetched and not **for the search** as it is used now, right?

**Comment by kpodemski** [comment]
@djfm yes, `hookActionProductListOverride` must be "fired" on *final* product list

**Comment by inetbiz** [comment]
@djfm too soon for https://schema.org/color as an example in the product-miniture or blocklayered until http://forge.prestashop.com/browse/PSCSX-5736 is addressed for various types of merchandise. Perhaps then blocklayered can output defined filter types according to schema.org vocabulary.

What I do mean for now is products on a category page should be in list item format. Reference the issue I posted to https://github.com/schemaorg/schemaorg/issues/605

```
<ul>
  <li>
    ... <! Product Miniature -->
  </li>
```
We need total count in that loop to populate itemlist number of items count. Either in JSON+LD or microformat.

**Comment by djfm** [comment]
@kpodemski but does `hookActionProductListOverride` really need to be able to change the products or could it be read-only? Maybe a `hookDisplayBeforeProductList` that would be called with the products about to be displayed would do the trick?

**Comment by kpodemski** [comment]
@djfm well, it should be read-write, right? because we need to modify outputted array of the products with some additional keys

**Comment by djfm** [comment]
@kpodemski but what do you do with the additional keys you add to the products? The template would need to know what to do with them :)

Would this useful to you? https://github.com/djfm/PrestaShop/commit/c2426ff10608424faa656d5abf2776978a007b04

Or do you think it should be a display hook whose output should be placed at the top of the product list?

**Comment by djfm** [comment]
See this PR instead: https://github.com/PrestaShop/PrestaShop/pull/4548

**Pull request closed without merging**